src/enumerate.rs

```diff
@@ -1 +1,202 @@
 //! State enumeration
+use crate::dfa;
+use crate::dfa::Dfa;
+use crate::dfa::State;
+use crate::eval;
+use crate::eval::RuntimeError;
+use crate::eval::Value;
+use crate::syntax::*;
+use std::collections::BTreeMap;
+
+type Env = BTreeMap<Id, Value>;
+
+/**
+ *
+ */
+pub fn enumerate(program: &Program, _input: &str) -> Result<Dfa<Symbol>, RuntimeError> {
+    // keep track of visited states and their environments
+    // let mut state_lookup: Map<State, Env> = Map::new();
+    let mut state_lookup: Map<State, Env> = Map::new();
+    // the transitions between states
+    let mut trans: Map<State, Map<Symbol, State>> = Map::new();
+    //state names
+    let mut names: Map<State, String> = Map::new();
+
+    let mut env_lookup: BTreeMap<Env, State> = BTreeMap::new();
+
+    //get initial state
+    let mut init_e = eval::init_env(program);
+    for stmt in &program.start {
+        eval::eval_stmt(stmt, &mut init_e, &program)?;
+    }
+
+    //add inital state
+    let init_s = dfa::State::fresh();
+    state_lookup.insert(init_s, init_e.clone());
+    env_lookup.insert(init_e.clone(), init_s);
+
+    let mut workqueue: Vec<State> = Vec::new();
```

Comment:
this is a worklist, not a workqueue (in your implementation, it is a stack but the stack part is a detail).